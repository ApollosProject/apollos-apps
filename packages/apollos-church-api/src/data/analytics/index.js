import resolver from './resolver';
import schema from './schema';
import dataSource from './data-source';

export { resolver, schema, dataSource };
