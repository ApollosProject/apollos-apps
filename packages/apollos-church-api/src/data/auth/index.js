import { gql } from 'apollo-server';
import { createGlobalId } from '../node';

// export { default as model } from './model';
export { default as dataSource } from './data-source';

export const schema = gql`
  type AuthenticatedUser {
    id: ID!
    profile: Person
  }

  type Authentication {
    user: AuthenticatedUser
    token: String
  }

  extend type Mutation {
    authenticate(identity: String!, password: String!): Authentication
    registerPerson(email: String!, password: String!): Authentication
  }

  extend type Query {
    currentUser: AuthenticatedUser
  }
`;

export const resolver = {
  Query: {
    currentUser: (root, args, { dataSources }) =>
      dataSources.Auth.getCurrentPerson(),
  },
  AuthenticatedUser: {
    id: ({ id }, args, context, { parentType }) =>
      createGlobalId(id, parentType.name),
    profile: (authUser) => authUser,
  },
  Authentication: {
    user: (root, args, { dataSources }) => dataSources.Auth.getCurrentPerson(),
  },
  Mutation: {
    authenticate: (root, { identity, password }, { dataSources }) =>
      dataSources.Auth.authenticate({ identity, password }),
    registerPerson: (root, args, { dataSources }) =>
      dataSources.Auth.registerPerson(args),
  },
};
