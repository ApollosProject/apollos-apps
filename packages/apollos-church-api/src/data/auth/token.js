import { AuthenticationError } from 'apollo-server';
import jwt from 'jsonwebtoken';

const secret = process.env.SECRET || 'ASea$2gadj#asd0';

export const parseToken = (token) => jwt.verify(token, secret);

export const registerToken = (token) => {
  try {
    const { cookie, sessionId } = parseToken(token);

    return {
      userToken: token,
      rockCookie: cookie,
      sessionId,
    };
  } catch (e) {
    throw new AuthenticationError('Invalid token');
  }
};

export const generateToken = (params) =>
  jwt.sign({ ...params }, secret, { expiresIn: '60d' });
