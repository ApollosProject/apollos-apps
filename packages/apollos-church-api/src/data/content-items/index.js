import { gql } from 'apollo-server';
import { get } from 'lodash';
import flow from 'lodash/fp/flow';
import omitBy from 'lodash/fp/omitBy';
import pickBy from 'lodash/fp/pickBy';
import mapValues from 'lodash/fp/mapValues';
import values from 'lodash/fp/values';
import sanitizeHtml from 'apollos-church-api/src/utils/sanitize-html';
import { Constants } from 'apollos-church-api/src/connectors/rock';
import { createGlobalId } from '../node';

const mapValuesWithKey = mapValues.convert({ cap: false });

// export { default as model } from './model';
export { default as dataSource } from './data-source';

export const schema = gql`
  type SharableContentItem implements Sharable {
    url: String
    message: String
    title: String
  }

  interface ContentItem {
    id: ID!
    title: String
    coverImage: ImageMedia
    images: [ImageMedia]
    videos: [VideoMedia]
    audios: [AudioMedia]
    htmlContent: String
    childContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    siblingContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    parentChannel: ContentChannel

    sharing: SharableContentItem
    theme: Theme
    isLiked: Boolean
  }

  type UniversalContentItem implements ContentItem & Node {
    id: ID!
    title: String
    coverImage: ImageMedia
    images: [ImageMedia]
    videos: [VideoMedia]
    audios: [AudioMedia]
    htmlContent: String
    childContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    siblingContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    parentChannel: ContentChannel
    terms(match: String): [Term]

    sharing: SharableContentItem
    theme: Theme
    isLiked: Boolean
  }

  type DevotionalContentItem implements ContentItem & Node {
    id: ID!
    title: String
    coverImage: ImageMedia
    images: [ImageMedia]
    videos: [VideoMedia]
    audios: [AudioMedia]
    htmlContent: String
    childContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    siblingContentItemsConnection(
      first: Int
      after: String
    ): ContentItemsConnection
    parentChannel: ContentChannel

    sharing: SharableContentItem
    theme: Theme
    isLiked: Boolean
    scriptures: Scripture
  }

  type Term {
    key: String
    value: String
  }

  input ContentItemsConnectionInput {
    first: Int
    after: String
  }

  type ContentItemsConnection {
    edges: [ContentItemsConnectionEdge]
    # TODO totalCount: Int
    # TODO pageInfo: PaginationInfo
  }

  type ContentItemsConnectionEdge {
    node: ContentItem
    cursor: String
  }

  extend type Query {
    userFeed(first: Int, after: String): ContentItemsConnection
    getAllLikedContent: [ContentItem]
  }
`;

const hasScripture = ({ attributeValues }) =>
  get(attributeValues, 'scriptures.value') != null;

const isImage = ({ key, attributeValues, attributes }) =>
  attributes[key].fieldTypeId === Constants.FIELD_TYPES.IMAGE ||
  (key.toLowerCase().includes('image') &&
    typeof attributeValues[key].value === 'string' &&
    attributeValues[key].value.startsWith('http')); // looks like an image url

const isVideo = ({ key, attributeValues, attributes }) =>
  attributes[key].fieldTypeId === Constants.FIELD_TYPES.VIDEO_FILE ||
  attributes[key].fieldTypeId === Constants.FIELD_TYPES.VIDEO_URL ||
  (key.toLowerCase().includes('video') &&
    typeof attributeValues[key].value === 'string' &&
    attributeValues[key].value.startsWith('http')); // looks like a video url

const isAudio = ({ key, attributeValues, attributes }) =>
  attributes[key].fieldTypeId === Constants.FIELD_TYPES.AUDIO_FILE ||
  attributes[key].fieldTypeId === Constants.FIELD_TYPES.AUDIO_URL ||
  (key.toLowerCase().includes('audio') &&
    typeof attributeValues[key].value === 'string' &&
    attributeValues[key].value.startsWith('http')); // looks like an audio url

export const defaultContentItemResolvers = {
  id: ({ id }, args, context, { parentType }) =>
    createGlobalId(id, parentType.name),
  htmlContent: ({ content }) => sanitizeHtml(content),
  childContentItemsConnection: async ({ id }, args, { dataSources }) =>
    dataSources.ContentItem.paginate({
      cursor: await dataSources.ContentItem.getCursorByParentContentItemId(id),
      args,
    }),

  parentChannel: ({ contentChannelId }, args, { dataSources }) =>
    dataSources.ContentChannel.getFromId(contentChannelId),

  siblingContentItemsConnection: async ({ id }, args, { dataSources }) =>
    dataSources.ContentItem.paginate({
      cursor: await dataSources.ContentItem.getCursorBySiblingContentItemId(id),
      args,
    }),

  images: ({ attributeValues, attributes }) => {
    const imageKeys = Object.keys(attributes).filter((key) =>
      isImage({
        key,
        attributeValues,
        attributes,
      })
    );
    return imageKeys.map((key) => ({
      __typename: 'ImageMedia',
      key,
      name: attributes[key].name,
      sources: attributeValues[key].value
        ? [{ uri: attributeValues[key].value }]
        : [],
    }));
  },

  videos: ({ attributeValues, attributes }) => {
    const videoKeys = Object.keys(attributes).filter((key) =>
      isVideo({
        key,
        attributeValues,
        attributes,
      })
    );
    return videoKeys.map((key) => ({
      __typename: 'VideoMedia',
      key,
      name: attributes[key].name,
      embedHtml: get(attributeValues, 'videoEmbed.value', null), // TODO: this assumes that the key `VideoEmebed` is always used on Rock
      sources: attributeValues[key].value
        ? [{ uri: attributeValues[key].value }]
        : [],
    }));
  },

  audios: ({ attributeValues, attributes }) => {
    const audioKeys = Object.keys(attributes).filter((key) =>
      isAudio({
        key,
        attributeValues,
        attributes,
      })
    );
    return audioKeys.map((key) => ({
      __typename: 'AudioMedia',
      key,
      name: attributes[key].name,
      sources: attributeValues[key].value
        ? [{ uri: attributeValues[key].value }]
        : [],
    }));
  },

  coverImage: async (root, args, { dataSources }) => {
    const pickBestImage = (images) => {
      // TODO: there's probably a _much_ more explicit and better way to handle this
      const squareImage = images.find((image) =>
        image.key.toLowerCase().includes('square')
      );
      if (squareImage) return { ...squareImage, __typename: 'ImageMedia' };
      return { ...images[0], __typename: 'ImageMedia' };
    };

    let defaultImages = defaultContentItemResolvers.images(root) || [];
    defaultImages = defaultImages.filter((image) => image.sources.length); // filter images w/o URLs
    if (defaultImages.length) return pickBestImage(defaultImages);

    // If no image, check parent for image:
    const parentItemsCursor = await dataSources.ContentItem.getCursorByChildContentItemId(
      root.id
    );
    if (!parentItemsCursor) return null;

    const parentItems = await parentItemsCursor.get();

    if (parentItems.length) {
      const parentImages = parentItems
        .map(defaultContentItemResolvers.images)
        .find((images) => images.length)
        .filter((image) => image.sources.length); // filter images w/o URLs

      if (parentImages && parentImages.length)
        return pickBestImage(parentImages);
    }

    return null;
  },

  // This resolver function is temporary, and is just used to get a seed to generate a random theme from
  theme: (root) => {
    if (![6, 5, 4].includes(root.contentChannelId)) return null; // todo: don't generate a theme for these content channel ids
    return root.guid; // todo: this `guid` is just being used as a seed to generate colors for now
  },
  isLiked: async ({ id, isLiked }, args, { dataSources }) => {
    if (isLiked != null) return isLiked;

    const interactions = await dataSources.Interactions.getForContentItem({
      contentItemId: id,
    });

    const likes = interactions.filter((i) => i.operation === 'Like').length;
    const unlike = interactions.filter((i) => i.operation === 'Unlike').length;
    // If likes / unlikes equal we have either unliked the content or haven't liked it yet (both are 0)
    return likes > unlike;
  },
  sharing: (root) => ({ __type: 'SharableContentItem', ...root }),
};

export const resolver = {
  Query: {
    userFeed: (root, args, { dataSources }) =>
      dataSources.ContentItem.paginate({
        cursor: dataSources.ContentItem.byUserFeed(),
        args,
      }),
    getAllLikedContent: async (root, args, { dataSources }) => {
      // Get All Interactions from current user
      const interactions = await dataSources.Interactions.getForContentItems();

      const likeCounts = {};

      // Iterate over the interactions and determine which pieces of content
      // has more likes than unlikes
      interactions.forEach(({ operation, relatedEntityId }) => {
        if (!likeCounts[relatedEntityId]) {
          likeCounts[relatedEntityId] = 0;
        }
        if (operation === 'Like') {
          likeCounts[relatedEntityId] += 1;
        }
        if (operation === 'Unlike') {
          likeCounts[relatedEntityId] -= 1;
        }
      });

      const itemIds = [];
      Object.keys(likeCounts).forEach((relatedEntityId) => {
        if (likeCounts[relatedEntityId] > 0) {
          itemIds.push(relatedEntityId);
        }
      });

      // Grab content related to user's interactions
      const getUserContentFromInteractions = itemIds.map((id) =>
        dataSources.ContentItem.getFromId(id)
      );

      const resolveUserContentFromInteractions = await Promise.all(
        getUserContentFromInteractions
      );

      // Determine the isLiked value on contentitems and create an obj that we
      // can merge with our main set of data later
      const calculateIsLikedOnContentItems = resolveUserContentFromInteractions.map(
        (item) => ({ ...item, isLiked: true })
      );

      return calculateIsLikedOnContentItems;
    },
  },
  DevotionalContentItem: {
    ...defaultContentItemResolvers,
    scriptures: ({ attributeValues }, args, { dataSources }) => {
      const reference = get(attributeValues, 'scriptures.value');
      if (reference && reference != null) {
        return dataSources.Scripture.getScripture(reference);
      }
      return null;
    },
  },
  UniversalContentItem: {
    ...defaultContentItemResolvers,
    terms: ({ attributeValues, attributes }, { match }) =>
      flow([
        omitBy((value, key) => isImage({ key, attributes, attributeValues })),
        omitBy((value, key) => isVideo({ key, attributes, attributeValues })),
        omitBy((value, key) => isAudio({ key, attributes, attributeValues })),
        omitBy((value, key) => key === 'videoEmbed'),
        pickBy((value, key) => (match ? key.match(match) : true)),
        mapValuesWithKey(({ value }, key) => ({
          key,
          value,
        })),
        values,
      ])(attributeValues),
  },
  ContentItem: {
    ...defaultContentItemResolvers,
    __resolveType: ({ attributeValues }) => {
      if (hasScripture({ attributeValues })) {
        return 'DevotionalContentItem';
      }
      return 'UniversalContentItem';
    },
  },
  SharableContentItem: {
    url: () => 'https://apollosrock.newspring.cc/', // todo: return a dynamic url that links to the content item
    title: ({ title }) => title,
    message: () => '',
  },
};
