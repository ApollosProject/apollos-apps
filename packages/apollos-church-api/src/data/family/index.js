import { schema, resolver } from './schema';
import dataSource from './data-source';

export { schema, resolver, dataSource };
