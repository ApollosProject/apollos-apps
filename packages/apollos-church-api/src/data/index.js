import { gql } from 'apollo-server';

import { createApolloServerConfig } from '@apollosproject/server-core';

import * as Analytics from '@apollosproject/data-connector-analytics';
import * as Scripture from '@apollosproject/data-connector-bible';
import * as LiveStream from '@apollosproject/data-connector-church-online';
import * as Cloudinary from '@apollosproject/data-connector-cloudinary';
import {
  Followings,
  Interactions,
  RockConstants,
  Family,
  Person,
  ContentItem,
  ContentChannel,
  Sharable,
  Auth,
} from '@apollosproject/data-connector-rock';
import * as Theme from './theme';

const data = {
  Followings,
  ContentChannel,
  ContentItem,
  Person,
  Cloudinary,
  Auth,
  LiveStream,
  Theme,
  Scripture,
  Interactions,
  RockConstants,
  Sharable,
  Analytics,
  Family,
  UniversalContentItem: {
    dataSource: ContentItem.dataSource,
  }, // alias
  DevotionalContentItem: {
    dataSource: ContentItem.dataSource,
  }, // alias
  ContentSeriesContentItem: {
    dataSource: ContentItem.dataSource,
  }, // alias
  MediaContentItem: {
    dataSource: ContentItem.dataSource,
  }, // alias
};

const { dataSources, resolvers, schema, context } = createApolloServerConfig(
  data
);

export { dataSources, resolvers, schema, context };

// the upload Scalar is added
export const testSchema = [
  gql`
    scalar Upload
  `,
  ...schema,
];
