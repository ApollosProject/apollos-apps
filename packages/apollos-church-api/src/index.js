import dotenv from 'dotenv/config'; // eslint-disable-line
import { ApolloServer } from 'apollo-server';
import { resolvers, schema, testSchema } from './data';

import getContext from './getContext';
import getDataSources from './getDataSources';

export { resolvers, schema, testSchema };

export default new ApolloServer({
  typeDefs: schema,
  resolvers,
  dataSources: getDataSources,
  context: getContext,
  formatError: (error) => {
    console.error(error.extensions.exception.stacktrace.join('\n'));
    return error;
  },
  playground: {
    settings: {
      'editor.cursorShape': 'line',
    },
  },
});
