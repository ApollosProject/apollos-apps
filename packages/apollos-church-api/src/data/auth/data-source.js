import { AuthenticationError } from 'apollo-server';
import { fetch, Request } from 'apollo-server-env';
import moment from 'moment';
import RockApolloDataSource from '@apollosproject/rock-apollo-data-source';

import { generateToken, registerToken } from './token';

export default class AuthDataSource extends RockApolloDataSource {
  resource = 'Auth';

  rockCookie = null;

  userToken = null;

  getCurrentPerson = async ({ cookie } = { cookie: null }) => {
    const { rockCookie } = this.context;
    const userCookie = cookie || rockCookie;
    if (userCookie) {
      const request = await this.request('People/GetCurrentPerson').get({
        options: { headers: { cookie: userCookie } },
      });
      return request;
    }
    throw new AuthenticationError('Must be logged in');
  };

  fetchUserCookie = async (Username, Password) => {
    try {
      // We use `new Response` rather than string/options b/c if conforms more closely with ApolloRESTDataSource
      // (makes mocking in tests WAY easier to use `new Request` as an input in both places)
      const response = await fetch(
        new Request(`${this.baseURL}/Auth/Login`, {
          method: 'POST',
          body: JSON.stringify({
            Username,
            Password,
          }),
          headers: {
            'Content-Type': 'Application/Json',
          },
        })
      );
      if (response.status >= 400) throw new AuthenticationError();
      const cookie = response.headers.get('set-cookie');
      return cookie;
    } catch (err) {
      throw new AuthenticationError('Invalid Credentials');
    }
  };

  createSession = async ({ cookie }) => {
    const currentUser = await this.getCurrentPerson({ cookie });
    return this.post('/InteractionSessions', {
      PersonAliasId: currentUser.primaryAliasId,
    });
  };

  authenticate = async ({ identity, password }) => {
    try {
      const cookie = await this.fetchUserCookie(identity, password);
      const sessionId = await this.createSession({ cookie });
      const token = generateToken({ cookie, sessionId });
      const { userToken, rockCookie } = registerToken(token);
      this.context.rockCookie = rockCookie;
      this.context.userToken = userToken;
      this.context.sessionId = sessionId;
      return { token, rockCookie };
    } catch (e) {
      throw e;
    }
  };

  personExists = async ({ identity }) => {
    const hasUserName = await this.request(
      `/UserLogins?$filter=UserName eq '${identity}'`
    ).get();

    if (hasUserName.length) {
      return true;
    }
    return false;
  };

  createUserProfile = async (props = {}) => {
    try {
      const { email } = props;

      return await this.post('/People', {
        Email: email,
        IsSystem: false, // Required by Rock
        Gender: 0, // Required by Rock
      });
    } catch (err) {
      throw new Error('Unable to create profile!');
    }
  };

  createUserLogin = async (props = {}) => {
    try {
      const { email, password, personId } = props;

      return await this.post('/UserLogins', {
        PersonId: personId,
        EntityTypeId: 27, // A default setting we use in Rock-person-creation-flow
        UserName: email,
        PlainTextPassword: password,
        LastLoginDateTime: `${moment().toISOString()}`,
      });
    } catch (err) {
      throw new Error('Unable to create user login!');
    }
  };

  changePassword = async ({ password }) => {
    const currentUser = await this.getCurrentPerson();
    const { email, id } = currentUser;
    const logins = await this.request('/UserLogins')
      .filter(`UserName eq '${email}'`)
      .get();

    if (logins.length > 0) {
      await this.delete(`/UserLogins/${logins[0].id}`);
    }
    await this.createUserLogin({
      personId: id,
      email,
      password,
    });
    return this.authenticate({
      identity: email,
      password,
    });
  };

  registerPerson = async ({ email, password }) => {
    const personExists = await this.personExists({ identity: email });
    if (personExists) throw new Error('User already exists!');

    const personId = await this.createUserProfile({ email });
    await this.createUserLogin({
      email,
      password,
      personId,
    });

    const token = await this.authenticate({ identity: email, password });
    return token;
  };
}
