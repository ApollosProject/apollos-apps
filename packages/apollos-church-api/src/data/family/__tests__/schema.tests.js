import { graphql } from 'graphql';
import { fetch } from 'apollo-server-env';
import { makeExecutableSchema } from 'apollo-server';

import { getTestContext } from 'apollos-church-api/src/utils/testUtils';
import { generateToken } from 'apollos-church-api/src/data/auth/token';
// we import the root-level schema and resolver so we test the entire integration:
import { testSchema as typeDefs, resolvers } from 'apollos-church-api/src/data';

describe('Person', () => {
  let schema;
  let context;
  beforeEach(() => {
    fetch.resetMocks();
    fetch.mockRockDataSourceAPI();
    schema = makeExecutableSchema({ typeDefs, resolvers });
    const token = generateToken({
      cookie: 'some-cookie',
      sessionId: 'somessessionid',
    });
    context = getTestContext({ req: { headers: { authorization: token } } });
  });

  it("returns a user's location", async () => {
    const query = `
      query {
        currentUser {
          profile {
            location
          }
        }
      }
    `;
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
});
