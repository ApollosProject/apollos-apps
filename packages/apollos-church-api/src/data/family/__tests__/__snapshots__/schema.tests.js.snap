// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person returns a user's location 1`] = `
Object {
  "data": Object {
    "currentUser": Object {
      "profile": Object {
        "location": "Chicago, IL",
      },
    },
  },
}
`;
